examples/adapter_discovery.py

```diff
@@ -0,0 +1,90 @@
+# SPDX-License-Identifier: LGPL-2.1-or-later
+# Copyright (C) 2022  Shin'ichiro Kawasaki <kawasaki@juno.dti.ne.jp>
+
+import sys
+import time
+import xml.etree.ElementTree as ET
+from asyncio import new_event_loop
+
+from sdbus import sd_bus_open_system
+from sdbus.dbus_proxy_async_interfaces import DbusIntrospectableAsync
+from sdbus_async.bluez.adapter_api import AdapterInterfaceAsync
+
+dbus = sd_bus_open_system()
+adapter = AdapterInterfaceAsync()
+adapter._connect('org.bluez', '/org/bluez/hci0', bus=dbus)
+adapter_introspect = DbusIntrospectableAsync()
+adapter_introspect._connect('org.bluez', '/org/bluez/hci0', bus=dbus)
+
+
+async def call_start_discovery() -> None:
+    try:
+        await adapter.start_discovery()
+    except NotImplementedError as e:
+        print(e)
+    except Exception as e:
+        print(e)
+        print(f"failed to start discovery: {e}")
+
+
+async def call_stop_discovery() -> None:
+    try:
+        await adapter.stop_discovery()
+    except NotImplementedError as e:
+        print(e)
+    except Exception as e:
+        print(e)
+        print(f"failed to stop discovery: {e}")
+
+
+async def call_introspect() -> None:
+    try:
+        s = await adapter_introspect.dbus_introspect()
+        parser = ET.fromstring(s)
+        nodes = parser.findall("./node")
+        if not nodes:
+            print("device not found")
+        else:
+            print(f"{len(nodes)} device(s) found")
+            for node in nodes:
+                print(f"  {node.attrib['name']}")
+    except NotImplementedError as e:
+        print(e)
+    except Exception as e:
+        print(e)
+        print(f"failed to stop discovery: {e}")
+
+
+def usage() -> None:
+    print(f"Usage: {sys.argv[0]} [OPTIONS]")
+    print("Call bluez adapter dicovery API and list found devices")
+    print("\tOPTIONS: -h:\t\tshow this help")
+    print("\t         -t TIME:\trun disvoery for TIME seconds")
+    exit(1)
+
+
+DISCOVERY_SECONDS = 5
+argv = sys.argv
+if len(argv) == 2 and argv[1] == '-h':
+    usage()
+elif len(argv) == 3 and argv[1] == '-t':
+    DISCOVERY_SECONDS = int(argv[2])
+elif len(argv) != 1:
+    usage()
+
+loop = new_event_loop()
+
+print("Starting discovery...")
+task = loop.create_task(call_start_discovery())
+loop.run_until_complete(task)
+
+print(f"Wait discovery for {DISCOVERY_SECONDS} seconds")
+time.sleep(DISCOVERY_SECONDS)
```

Comment:
Mixing blocking sleep and asyncio is kind of a bad practice. It will block all sdbus processing and you might get disconnected from D-Bus. There is `asyncio.sleep` with similar functionality but allows other async code to execute in parallel.

Reply:
Thanks for the advice. I replaced the time.sleep with asyncio.sleep and moved to call_start_discovery().